### PR TITLE
allow querying IPMI sensors via the open interface

### LINKED ISSUE #1547: ipmi_sensor support connecting locally rather than via TCP
Dell R730:

```
[root@host ~]# ipmitool sdr type Temperature
Inlet Temp       | 04h | ok  |  7.1 | 23 degrees C
Exhaust Temp     | 01h | ok  |  7.1 | 30 degrees C
Temp             | 0Eh | ok  |  3.1 | 43 degrees C
Temp             | 0Fh | ok  |  3.2 | 45 degrees C
```

Dell R630:

```
[root@host ~]# ipmitool sdr type Temperature
Inlet Temp       | 04h | ok  |  7.1 | 23 degrees C
Exhaust Temp     | 01h | ok  |  7.1 | 33 degrees C
Temp             | 0Eh | ok  |  3.1 | 50 degrees C
Temp             | 0Fh | ok  |  3.2 | 57 degrees C
```

Supermicro based board: 

```
[root@host ~]# ipmitool sdr type Temperature
SSB Therm Trip   | 0Dh | ok  |  7.1 |
BB Inlet Temp    | 20h | ok  |  7.1 | 29 degrees C
SSB Temp         | 22h | ok  |  7.1 | 60 degrees C
BB BMC Temp      | 23h | ok  |  7.1 | 50 degrees C
P1 VR Temp       | 24h | ok  |  7.1 | 41 degrees C
IB Temp          | 25h | ok  |  7.1 | 44 degrees C
LAN NIC Temp     | 2Fh | ok  | 12.2 | 64 degrees C
P1 Therm Margin  | 74h | ok  |  3.1 | -21 degrees C
P2 Therm Margin  | 75h | ok  |  3.2 | -36 degrees C
P1 Therm Ctrl %  | 78h | ok  |  3.1 | 0 percent
P2 Therm Ctrl %  | 79h | ok  |  3.2 | 0 percent
P1 DTS Therm Mgn | 83h | ok  |  3.1 | -21 degrees C
P2 DTS Therm Mgn | 84h | ok  |  3.2 | -36 degrees C
P1 VRD Hot       | 90h | ok  |  3.1 |
P2 VRD Hot       | 91h | ok  |  3.2 |
DIMM Thrm Mrgn 1 | B0h | ok  |  7.1 | -45 degrees C
DIMM Thrm Mrgn 2 | B1h | ok  |  7.1 | -46 degrees C
DIMM Thrm Mrgn 3 | B2h | ok  |  7.1 | -56 degrees C
DIMM Thrm Mrgn 4 | B3h | ok  |  7.1 | -54 degrees C
```

HP ProLiant DL360p Gen8 SE:

```
[root@host ~]# ipmitool sdr type Temperature
01-Inlet Ambient | 03h | ok  | 64.1 | 21 degrees C
02-CPU 1         | 04h | ok  | 65.1 | 40 degrees C
03-CPU 2         | 05h | ok  | 65.2 | 40 degrees C
04-P1 DIMM 1-6   | 06h | ok  | 32.1 | 31 degrees C
05-P1 DIMM 7-12  | 07h | ok  | 32.2 | 31 degrees C
06-P2 DIMM 1-6   | 08h | ok  | 32.3 | 31 degrees C
07-P2 DIMM 7-12  | 09h | ok  | 32.4 | 31 degrees C
08-P1 Mem Zone   | 0Ah | ok  | 32.5 | 27 degrees C
09-P1 Mem Zone   | 0Bh | ok  | 32.6 | 30 degrees C
10-P2 Mem Zone   | 0Ch | ok  | 32.7 | 30 degrees C
11-P2 Mem Zone   | 0Dh | ok  | 32.8 | 27 degrees C
12-HD Max        | 0Eh | ok  |  4.1 | 35 degrees C
13-Chipset 1     | 0Fh | ok  | 66.1 | 44 degrees C
14-Chipset1 Zone | 10h | ok  | 66.2 | 31 degrees C
15-P/S 1 Inlet   | 11h | ok  | 10.4 | 27 degrees C
16-P/S 1 Zone    | 12h | ok  | 10.5 | 30 degrees C
17-P/S 2 Inlet   | 13h | ok  | 10.6 | 30 degrees C
18-P/S 2 Zone    | 14h | ok  | 10.7 | 31 degrees C
19-PCI #1        | 15h | ns  | 11.1 | Disabled
20-PCI #2        | 16h | ns  | 11.2 | Disabled
21-VR P1         | 17h | ok  | 19.1 | 34 degrees C
22-VR P2         | 18h | ok  | 19.2 | 35 degrees C
23-VR P1 Mem     | 19h | ok  | 19.3 | 26 degrees C
24-VR P1 Mem     | 1Ah | ok  | 19.4 | 25 degrees C
25-VR P2 Mem     | 1Bh | ok  | 19.5 | 27 degrees C
26-VR P2 Mem     | 1Ch | ok  | 19.6 | 26 degrees C
27-VR P1Mem Zone | 1Dh | ok  | 19.7 | 25 degrees C
28-VR P1Mem Zone | 1Eh | ok  | 19.8 | 26 degrees C
29-VR P2Mem Zone | 1Fh | ok  | 19.9 | 27 degrees C
30-VR P2Mem Zone | 20h | ok  | 19.10 | 25 degrees C
31-HD Controller | 21h | ok  | 66.3 | 52 degrees C
32-HD Cntlr Zone | 22h | ok  | 66.4 | 35 degrees C
33-PCI 1 Zone    | 23h | ok  | 66.5 | 30 degrees C
34-PCI 1 Zone    | 24h | ok  | 66.6 | 31 degrees C
35-LOM Card      | 25h | ns  | 11.3 | Disabled
36-PCI 2 Zone    | 26h | ok  | 66.7 | 33 degrees C
37-System Board  | 27h | ok  | 66.8 | 32 degrees C
38-System Board  | 28h | ok  | 66.9 | 31 degrees C
39-Sys Exhaust   | 29h | ok  | 66.10 | 30 degrees C
40-Sys Exhaust   | 2Ah | ok  | 66.11 | 31 degrees C
41-Sys Exhaust   | 2Bh | ok  | 66.12 | 31 degrees C
42-SuperCAP Max  | 2Ch | ok  | 40.1 | 24 degrees C
```

SGI C2110G-RP5 (also supermicro based):

```
[root@host ~]# ipmitool sdr type Temperature
CPU1 Temp        | 01h | ok  |  3.1 | 36 degrees C
CPU2 Temp        | 02h | ok  |  3.2 | 33 degrees C
System Temp      | 11h | ok  |  7.1 | 23 degrees C
Peripheral Temp  | 12h | ok  |  7.2 | 30 degrees C
PCH Temp         | 0Ah | ok  |  7.3 | 45 degrees C
P1-DIMMA TEMP    | B0h | ok  | 32.64 | 33 degrees C
P1-DIMMB TEMP    | B4h | ok  | 32.68 | 33 degrees C
P1-DIMMC TEMP    | B8h | ok  | 32.72 | 35 degrees C
P1-DIMMD TEMP    | BCh | ok  | 32.76 | 37 degrees C
P2-DIMME TEMP    | D0h | ok  | 32.80 | 29 degrees C
P2-DIMMF TEMP    | D4h | ok  | 32.84 | 31 degrees C
P2-DIMMG TEMP    | D8h | ok  | 32.88 | 28 degrees C
P2-DIMMH TEMP    | DCh | ok  | 32.92 | 31 degrees C
GPU1 Temp        | 71h | ok  | 11.1 | 24 degrees C
GPU2 Temp        | 72h | ok  | 11.2 | 21 degrees C
GPU3 Temp        | 73h | ok  | 11.3 | 21 degrees C
GPU4 Temp        | 74h | ok  | 11.4 | 21 degrees C
GPU5 Temp        | 75h | ns  | 11.5 | No Reading
GPU6 Temp        | 76h | ns  | 11.6 | No Reading
Xeon Phi1 Temp   | 79h | ns  | 11.9 | No Reading
Xeon Phi2 Temp   | 7Ah | ns  | 11.10 | No Reading
Xeon Phi3 Temp   | 7Bh | ns  | 11.11 | No Reading
Xeon Phi4 Temp   | 7Ch | ns  | 11.12 | No Reading
Xeon Phi5 Temp   | 7Dh | ns  | 11.13 | No Reading
Xeon Phi6 Temp   | 7Eh | ns  | 11.14 | No Reading
```

I think you get the idea. Pretty much every vendor provide some set of sensor data in degrees C.

There are other types of sensor data other than "sdr type Temperature", although we have no interest in capturing them. 

We propose creating a ipmi plugin that
- Initially, captures all the counters inside "sdr type Temperature" (in degrees C). Those marked "No Reading", "Disabled", etc. will just be skipped.
- Uses a naming format that would allow other types of sensor to be added in future
- Casts the field names from the output to lowercase, replacing spaces with underscores
- Handles ipmitool returning extremely slowly with timeouts (this happens).

RFC. We are happy to work on this and submit a PR.

### PR DESCRIPTION
**Pull request by m4ce**
Hi,

this solves #1547.

I have not adjuted the tests yet. I wanted to first confirm with you that my changes are okay and then update the tests accordingly.

Cheers,
Matteo

### CODE REVIEW

**Comment by m4ce** [comment]
I've also updated the README so that you can evaluate the changes better.

**Comment by m4ce** [comment]
@sparrc, any change of having this in the next release? :(

**Comment by m4ce** [comment]
All tests are passing now, I've re-factored the code a little bit more.

File: plugins/inputs/ipmi_sensor/README.md [hunk 1]
```diff
@@ -4,23 +4,32 @@ Get bare metal metrics using the command line utility `ipmitool`
 
 see ipmitool(https://sourceforge.net/projects/ipmitool/files/ipmitool/)
 
-The plugin will use the following command to collect remote host sensor stats:
+The plugin supports two modes. When 'servers' is not specified, the plugin will utilize the OpenIPMI kernel device driver to retrieve the sensor stats.
```
Comment (sparrc) [inline plugins/inputs/ipmi_sensor/README.md]: what does this mean? does it still run an ipmitool command? can you provide it here?

File: plugins/inputs/ipmi_sensor/README.md [hunk 2]
```diff
@@ -4,23 +4,32 @@ Get bare metal metrics using the command line utility `ipmitool`
...
     * Fields:
       - status
       - value
 
+The `server` tag will be made available when retrieving stats from remote server(s).
+
 ## Configuration
 
 ```toml
 [[inputs.ipmi_sensor]]
```
Comment (sparrc) [inline plugins/inputs/ipmi_sensor/README.md]: you should change the SampleConfig to indicate that `servers` is optional

File: plugins/inputs/ipmi_sensor/ipmi.go [hunk 3]
```diff
@@ -23,26 +31,28 @@ var sampleConfig = `
...
-	}
-}
-
 func (m *Ipmi) SampleConfig() string {
 	return sampleConfig
 }
 
 func (m *Ipmi) Description() string {
-	return "Read metrics from one or many bare metal servers"
+	return "Read metrics from the local machines (via the open interface) or one or many bare metal servers (via the lan interface)"
```
Comment (sparrc) [inline plugins/inputs/ipmi_sensor/ipmi.go]: limit this to 80 characters

File: plugins/inputs/ipmi_sensor/ipmi.go [hunk 4]
```diff
@@ -23,26 +31,28 @@ var sampleConfig = `
...
 }
 
 func (m *Ipmi) Gather(acc telegraf.Accumulator) error {
-	if m.runner == nil {
-		m.runner = CommandRunner{}
+	if len(m.path) == 0 {
+		return errors.New("ipmitool not found: verify that ipmitool is installed and that ipmitool is in your PATH")
```
Comment (sparrc) [inline plugins/inputs/ipmi_sensor/ipmi.go]: use fmt.Errorf instead of errors.New

**Comment by sparrc** [comment]
@m4ce I'm not sure I understand the distinction you're making in documentation with local vs. external hosts using the "OpenIPMI device driver"? Don't they both use OpenIPMI?

**Comment by m4ce** [comment]
@sparrc, when use run ```ipmitool sdr``` without specifying the interface (-I), it uses the open interface by default (/dev/ipmi<X>). The open interface uses the OpenIPMI kernel device driver to query the local machine sensors. When you use 'lan' as an interface, you are expected to pass a host via the -H command line flag. It still uses the openipmi protocol, but over TCP.

**Comment by sparrc** [comment]
but in both cases we are querying the OpenIPMI device driver, it's just that in the `lan` case we are querying a remote server.

I think you should change the documentation to only specify that if a server is not given, then we will query locally.

**Comment by m4ce** [comment]
@sparrc, I've improved the documentation and amended the code based on your reviews.

File: plugins/inputs/ipmi_sensor/ipmi.go [hunk 5]
```diff
@@ -1,48 +1,61 @@
...
 	Servers []string
-	runner  Runner
 }
 
 var sampleConfig = `
-  ## specify servers via a url matching:
+  ## optionally specify the path to the ipmitool executable
+  ## path = "/usr/bin/ipmitool"
+
+  ## optionally specify one or more servers via a url matching. If servers is omitted, only the local machine sensor stats will be retrieved.
```
Comment (sparrc) [inline plugins/inputs/ipmi_sensor/ipmi.go]: config file line length should be limited to 80 characters

**Comment by sparrc** [comment]
thanks @m4ce

**Pull request closed without merging**